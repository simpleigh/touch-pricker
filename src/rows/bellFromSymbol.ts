/**
 * Free Touch Pricker
 * @author Leigh Simpson <code@simpleigh.com>
 * @license GPL-3.0
 * @copyright Copyright 2015-23 Leigh Simpson. All rights reserved.
 */

import { Bell } from './types';

/* eslint-disable
@typescript-eslint/naming-convention,
sort-keys,
*/
// prettier-ignore
const BELL_SYMBOLS_MAP: Partial<Record<string, Bell>> = {
    '1': 1, '2': 2, '3': 3, '4': 4,
    '5': 5, '6': 6, '7': 7, '8': 8,
    '9': 9, '0': 10, 'E': 11, 'T': 12,
    'A': 13, 'B': 14, 'C': 15, 'D': 16,
};
/* eslint-enable */

/**
 * Converts a character into a {@link Bell}.
 *
 * Tries to convert an individual bell character into a bell number.
 * An exception is thrown if the character cannot be recognised.
 *
 * ```
 * > Pricker.bellFromSymbol('E');
 * 11
 * ```
 */
const bellFromSymbol = (input: string): Bell => {
    const upper = input.toUpperCase();

    if (!BELL_SYMBOLS_MAP[upper]) {
        throw new Error(`Unknown bell '${input}'`);
    }

    return BELL_SYMBOLS_MAP[upper]!;
};

export default bellFromSymbol;
