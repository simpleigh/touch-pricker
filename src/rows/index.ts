/**
 * Free Touch Pricker
 * @author Leigh Simpson <code@simpleigh.com>
 * @license GPL-3.0
 * @copyright Copyright 2015-23 Leigh Simpson. All rights reserved.
 */

export { default as bellFromSymbol } from './bellFromSymbol';
export { default as changeFromNotation } from './changeFromNotation';
export { default as inverse } from './inverse';
export { default as multiply } from './multiply';
export { default as rankFromRow } from './rankFromRow';
export { default as rounds } from './rounds';
export { default as rowFromRank } from './rowFromRank';
export { default as rowFromString } from './rowFromString';
export { default as stringFromRow } from './stringFromRow';
export { default as symbolFromBell } from './symbolFromBell';
export { Stage } from './types';
export type { Bell, Change, MutableRow, Row } from './types';
