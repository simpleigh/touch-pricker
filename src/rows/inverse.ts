/**
 * Free Touch Pricker
 * @author Leigh Simpson <code@simpleigh.com>
 * @license GPL-3.0
 * @copyright Copyright 2015-23 Leigh Simpson. All rights reserved.
 */

import { Bell, MutableRow, Row } from './types';

/**
 * Computes the inverse of a row.
 *
 * From Knuth 1.3.3:
 *
 * > The inverse π<sup>-</sup> of a permutation π is the rearrangement that
 * > undoes the effect of π; if _i_ goes to _j_ under π, then _j_ goes to _i_
 * > under π<sup>-</sup>. Thus the product ππ<sup>-</sup> equals the identity
 * > permutation, and so does the product π<sup>-</sup>π.
 *
 * ```
 * > inverse([5, 3, 1, 2, 4, 6])
 * [3, 4, 2, 5, 1, 6]
 * > multiply([5, 3, 1, 2, 4, 6], inverse([5, 3, 1, 2, 4, 6]))
 * [1, 2, 3, 4, 5, 6]
 * > multiply(inverse([5, 3, 1, 2, 4, 6]), [5, 3, 1, 2, 4, 6])
 * [1, 2, 3, 4, 5, 6]
 * ```
 *
 * A new row is returned; the original is left unchanged.
 */
const inverse = (row: Row): Row => {
    const output: MutableRow = [];

    for (let i = 0; i < row.length; i += 1) {
        output[row[i] - 1] = (i + 1) as Bell;
    }

    return output;
};

export default inverse;
