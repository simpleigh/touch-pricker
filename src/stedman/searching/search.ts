/**
 * Free Touch Pricker
 * @author Leigh Simpson <code@simpleigh.com>
 * @license GPL-3.0
 * @copyright Copyright 2015-23 Leigh Simpson. All rights reserved.
 * The function in this file is inspired by the C++ implementation of StedTurn
 * written by Philip Saddleton.
 */

import {
    multiply,
    rankFromRow,
    Row,
    rowFromRank,
    Uint4Table,
} from '../../rows';
import CallPair from './CallPair';
import createTranspositions from './createTranspositions';

const recursiveSearch = (
    table: Uint4Table,
    targetRank: number,
    steps: number,
    transpositions: Record<CallPair, Row>,
): string[] => {
    // Halt recursion if we've run out of steps.
    if (steps === 0) {
        return ['']; // base case: a single touch of zero length
    }

    const touches: string[] = [];

    // Loop through each possible calling.
    for (const calling of Object.getOwnPropertyNames(transpositions)) {
        const oldRow = rowFromRank(targetRank, table.stage);
        const newRow = multiply(oldRow, transpositions[calling as CallPair]);
        const newRank = rankFromRow(newRow);

        if (table.getValue(newRank) < steps) {
            // If we've got closer to the target then recurse.
            const childTouches = recursiveSearch(
                table,
                newRank,
                steps - 1,
                transpositions,
            );
            for (const touch of childTouches) {
                // Prune touches with undesirable calling.
                const touchEndsWithCall =
                    touch.length && touch[touch.length - 1] !== ' ';
                const callStartsWithSingle = calling[0] === 's';

                if (!(touchEndsWithCall && callStartsWithSingle)) {
                    touches.push(`${touch}${calling}`);
                }
            }
        }
    }

    return touches;
};

/**
 * Searches for all callings from rounds to a target row.
 *
 * We search by exploring a tree structure where each node in the tree
 * represents a single {@link Row}. We work backwards from the target row hoping
 * to find rounds at the required number of steps.
 *
 * Sixes (in Stedman) alternate between slow and quick and we can work around
 * this complexity by searching in steps of a "pair" of sixes, being a call, a
 * slow six, another call, and a quick six. Any row can be followed by any of
 * seven other rows: one per {@link CallPair}.
 *
 * The tree therefore expands rapidly. The node for the target row is connected
 * to seven other nodes; the next layer has 49 nodes and there's 343 after that.
 * By the time we get to 11 steps (22 sixes or a course of Stedman Cinques)
 * we would have 2,306,881,200 nodes in our tree. We need to optimise.
 *
 * We can prune the tree as we go by using a data table that stores the minimum
 * number of steps needed to get from rounds to the target row. This can easily
 * be generated in advance and works as follows:
 *
 * 1. Say (for example) our target row can only be reached from rounds after a
 *   minimum of 7 steps.
 * 2. Iterate over the seven nodes connected to the target row (representing the
 *   rows two sixes before the target row with each {@link CallPair}).
 * 3. For each node, check the minimum distance from rounds. If that is more
 *   than 6 steps then discard the node (we're moving away from rounds).
 * 4. Continue on from each remaining node.
 *
 * We've already filtered the list of {@link CallPair}s to exclude undesirable
 * callings (being `-s` or `ss`). We also need to filter out touches where these
 * callings develop _between_ pairs. We can do this as we go.
 * @param table  Data table being used for the search.
 * @param targetRank  Row we're trying to reach from rounds.
 * @param steps  Number of steps we can take to get to rounds.
 */
const search = (
    table: Uint4Table,
    targetRank: number,
    steps?: number,
): string[] => {
    const transpositions = createTranspositions(table.stage, true);
    steps ??= table.getValue(targetRank);
    return recursiveSearch(table, targetRank, steps, transpositions);
};

export default search;
