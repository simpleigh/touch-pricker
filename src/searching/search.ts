/**
 * Free Touch Pricker
 * @author Leigh Simpson <code@simpleigh.com>
 * @license GPL-3.0
 * @copyright Copyright 2015-23 Leigh Simpson. All rights reserved.
 * The algorithm in this file is adapted from the C++ implementation of StedTurn
 * written by Philip Saddleton.
 */

import { multiply, rankFromRow, type Row, rowFromRank } from '../rows';
import Calling from './Calling';
import type Uint4Table from './Uint4Table';

/**
 * A cache of search results.
 *
 * It's common while searching to visit the same node multiple times, i.e. to
 * reach the same six end in lots of different ways. Recalculating following
 * nodes repeatedly is wasteful so we use a cache to store the calculation for
 * immediate reuse if needed.
 *
 * We need to store different results for different numbers of steps so the
 * overall cache structure might look something like:
 *
 * ```
 * {
 *     0: { // rounds
 *         0: [''], // callings with zero steps
 *     },
 *     249: {
 *         4: ['- -- s  ', '-  -   s', '- s- s s'],
 *     },
 *     442: {
 *         2: ['- s-'],
 *     },
 *     551: {
 *         3: ['-  -  ', '- s- s'],
 *     },
 *     // ...
 * }
 * ```
 */
type Cache = Partial<Record<number, Partial<Record<number, string[]>>>>;

/**
 * Extend each of a list of callings.
 *
 * While searching for touches we need to append the current calling being
 * examined onto each of a list of child callings returned recursively.
 * This also provides an opportunity to filter out undesirable callings.
 * @param touchList  Array of child touches produced by a recursive call
 * @param calling  Additional call to append at the end of each child touch
 */
export const extendTouchList = (
    touchList: string[],
    calling: string,
): string[] => {
    // Filter the list to avoid undesirable callings like `-s` or `ss`.
    if (calling.startsWith('s')) {
        touchList = touchList.filter(
            (touch) => touch === '' || touch.endsWith(' '),
        );
    }

    return touchList.map((touch) => `${touch}${calling}`);
};

/**
 * The search implementation itself.
 *
 * See {@link search} for a fuller explanation of this algorithm.
 */
const recursiveSearch = (
    table: Uint4Table,
    transpositions: Map<string, Row>,
    targetRank: number,
    steps: number,
    cache: Cache = {},
): string[] => {
    // Halt recursion if we've run out of steps.
    if (steps === 0) {
        return ['']; // base case: a single touch of zero length
    }

    // If we already have a result in the cache then return it.
    if (cache[targetRank]?.[steps]) {
        return cache[targetRank]![steps]!;
    }

    let touches: string[] = [];

    // Loop through each possible calling.
    for (const [calling, transposition] of transpositions.entries()) {
        const oldRow = rowFromRank(targetRank, table.stage);
        const newRow = multiply(oldRow, transposition);
        const newRank = rankFromRow(newRow);

        if (table.getValue(newRank) < steps) {
            // If we've got closer to the target then recurse.
            const childTouches = recursiveSearch(
                table,
                transpositions,
                newRank,
                steps - 1,
                cache,
            );

            // Store the result in the cache.
            cache[newRank] ??= {};
            cache[newRank]![steps - 1] = childTouches;

            touches = [...touches, ...extendTouchList(childTouches, calling)];
        }
    }

    return touches;
};

/**
 * Searches for all callings from rounds to a target row.
 *
 * We search by exploring a tree structure where each node in the tree
 * represents a single {@link Row}. We work backwards from the target row hoping
 * to find rounds at the required number of steps.
 *
 * Sixes (in Stedman) alternate between slow and quick and we can work around
 * this complexity by searching in steps of a "pair" of sixes, being a call, a
 * slow six, another call, and a quick six. Any row can be followed by any of
 * seven other rows.
 *
 * The tree therefore expands rapidly. The node for the target row is connected
 * to seven other nodes; the next layer has 49 nodes and there's 343 after that.
 * By the time we get to 11 steps (22 sixes or a course of Stedman Cinques)
 * we would have 2,306,881,200 nodes in our tree. We need to optimise.
 *
 * We can prune the tree as we go by using a data table that stores the minimum
 * number of steps needed to get from rounds to the target row. This can easily
 * be generated in advance and works as follows:
 *
 * 1. Say (for example) our target row can only be reached from rounds after a
 *   minimum of 7 steps.
 * 2. Iterate over the seven nodes connected to the target row (representing the
 *   rows two sixes before the target row with each calling transposition).
 * 3. For each node, check the minimum distance from rounds. If that is more
 *   than 6 steps then discard the node (we're moving away from rounds).
 * 4. Continue on from each remaining node.
 *
 * Our list of transpositions already excludes undesirable callings (being `-s`
 * or `ss`). We also need to filter out touches where these callings develop
 * _between_ pairs. We can do this as we go.
 * @param table  Data table being used for the search.
 * @param transpositions  Transpositions to use to move between nodes.
 * @param targetRank  Row we're trying to reach from rounds.
 * @param steps  Number of steps we can take to get to rounds.
 */
const search = (
    table: Uint4Table,
    transpositions: Map<string, Row>,
    targetRank: number,
    steps?: number,
): Calling[] => {
    steps ??= table.getValue(targetRank);
    return recursiveSearch(table, transpositions, targetRank, steps).map(
        (calling: string): Calling => new Calling(calling),
    );
};

/**
 * Asynchronous version of {@link search}.
 *
 * Returns a `Promise` that resolves when the search is complete. Can be used to
 * yield control to the browser so that it can complete any DOM changes before
 * starting a search.
 */
export const searchAsync = async (
    table: Uint4Table,
    transpositions: Map<string, Row>,
    targetRank: number,
    steps?: number,
): Promise<Calling[]> =>
    new Promise<Calling[]>((resolve) => {
        setTimeout(() => {
            resolve(search(table, transpositions, targetRank, steps));
        });
    });

export default search;
