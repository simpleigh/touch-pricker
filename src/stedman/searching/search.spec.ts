/**
 * Free Touch Pricker
 * @author Leigh Simpson <code@simpleigh.com>
 * @license GPL-3.0
 * @copyright Copyright 2015-23 Leigh Simpson. All rights reserved.
 */

import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { Stage, Uint4Table } from '../../rows';
import search from './search';

describe('search function', () => {
    let table: Uint4Table;

    beforeAll(async () => {
        const filename = path.join(__dirname, '../../../data/stedman.7.dat');
        const buffer = await readFile(filename);
        const data = new Uint8Array(buffer.buffer);
        table = new Uint4Table(Stage.Triples, data);
    });

    it('can find touches', () => {
        const touches = search(table, 5039);
        expect(touches.length).toBe(6);
    });

    it('finds the expected touches', () => {
        const touches = search(table, 5039);
        expect(touches[0].print('text')).toBe('1 s3 4 6 8 10 14');
        expect(touches[1].print('text')).toBe('1 4 8 9 s12 14');
        expect(touches[2].print('text')).toBe('1 s3 4 s6 8 9 s12 14');
        expect(touches[3].print('text')).toBe('1 3 4 s6 s9 s12 14');
        expect(touches[4].print('text')).toBe('1 3 4 5 6 9 10 s12 14');
        expect(touches[5].print('text')).toBe('1 s3 4 6 s8 9 10 s12 14');
    });

    it('can find a large volume of touches', () => {
        const touches = search(table, 39);
        expect(touches.length).toBe(2639);
    });

    it('can find touches taking more sixes than the minimum', () => {
        const touches = search(table, 5039, 8);
        expect(touches.length).toBe(300);
        expect(touches).toMatchSnapshot();
    });
});
