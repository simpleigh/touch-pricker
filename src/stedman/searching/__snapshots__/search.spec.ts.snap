// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search function can find touches taking more sixes than the minimum 1`] = `
[
  Calling {
    "calling": "   s-- --  -    ",
  },
  Calling {
    "calling": "   - -  s  -    ",
  },
  Calling {
    "calling": "   s -  -- -    ",
  },
  Calling {
    "calling": "  ---  s-- -    ",
  },
  Calling {
    "calling": "s - s  s-- -    ",
  },
  Calling {
    "calling": "   --- s-- -    ",
  },
  Calling {
    "calling": "   s-  - -  -   ",
  },
  Calling {
    "calling": "s - -- - -  -   ",
  },
  Calling {
    "calling": "  s--- - -  -   ",
  },
  Calling {
    "calling": "s ---- --  --   ",
  },
  Calling {
    "calling": "  - s- --  --   ",
  },
  Calling {
    "calling": "  -  -  s  --   ",
  },
  Calling {
    "calling": "s s- -  s  --   ",
  },
  Calling {
    "calling": "-  s-  - - --   ",
  },
  Calling {
    "calling": "- s--- - - --   ",
  },
  Calling {
    "calling": "s -- -  -- --   ",
  },
  Calling {
    "calling": "s  s-  s-- --   ",
  },
  Calling {
    "calling": "  - -- s-- --   ",
  },
  Calling {
    "calling": "s s--- s-- --   ",
  },
  Calling {
    "calling": "  ---- --  s-   ",
  },
  Calling {
    "calling": "s - s- --  s-   ",
  },
  Calling {
    "calling": "s -  -  s  s-   ",
  },
  Calling {
    "calling": "  s- -  s  s-   ",
  },
  Calling {
    "calling": "  -- -  -- s-   ",
  },
  Calling {
    "calling": "   s-  s-- s-   ",
  },
  Calling {
    "calling": "s - -- s-- s-   ",
  },
  Calling {
    "calling": "  s--- s-- s-   ",
  },
  Calling {
    "calling": "- ---- --   s   ",
  },
  Calling {
    "calling": "- s- -  s   s   ",
  },
  Calling {
    "calling": "s  s-  - -  s   ",
  },
  Calling {
    "calling": "  - -- - -  s   ",
  },
  Calling {
    "calling": "s s--- - -  s   ",
  },
  Calling {
    "calling": "- -- -  --  s   ",
  },
  Calling {
    "calling": "-  s-  s--  s   ",
  },
  Calling {
    "calling": "- s--- s--  s   ",
  },
  Calling {
    "calling": "   - -    -  s  ",
  },
  Calling {
    "calling": "   s-- s  -  s  ",
  },
  Calling {
    "calling": "   s -   s-  s  ",
  },
  Calling {
    "calling": "  ---  s s-  s  ",
  },
  Calling {
    "calling": "s - s  s s-  s  ",
  },
  Calling {
    "calling": "   --- s s-  s  ",
  },
  Calling {
    "calling": "   s s- - -- s  ",
  },
  Calling {
    "calling": "  --- --- -- s  ",
  },
  Calling {
    "calling": "s - s --- -- s  ",
  },
  Calling {
    "calling": "   ------ -- s  ",
  },
  Calling {
    "calling": "s -   - s -- s  ",
  },
  Calling {
    "calling": "  s-  - s -- s  ",
  },
  Calling {
    "calling": "   - s- s -- s  ",
  },
  Calling {
    "calling": "  ---  ----- s  ",
  },
  Calling {
    "calling": "s - s  ----- s  ",
  },
  Calling {
    "calling": "   --- ----- s  ",
  },
  Calling {
    "calling": "   s -  s--- s  ",
  },
  Calling {
    "calling": "   s-- -- s- s  ",
  },
  Calling {
    "calling": "   - -  s s- s  ",
  },
  Calling {
    "calling": "- s- -      --  ",
  },
  Calling {
    "calling": "- ---- s    --  ",
  },
  Calling {
    "calling": "s -- -   -  --  ",
  },
  Calling {
    "calling": "s  s-  s -  --  ",
  },
  Calling {
    "calling": "  - -- s -  --  ",
  },
  Calling {
    "calling": "s s--- s -  --  ",
  },
  Calling {
    "calling": "- -- -   s  --  ",
  },
  Calling {
    "calling": "-  s-  s s  --  ",
  },
  Calling {
    "calling": "- s--- s s  --  ",
  },
  Calling {
    "calling": "s -  -    - --  ",
  },
  Calling {
    "calling": "  s- -    - --  ",
  },
  Calling {
    "calling": "  ---- s  - --  ",
  },
  Calling {
    "calling": "s - s- s  - --  ",
  },
  Calling {
    "calling": "  -- -   s- --  ",
  },
  Calling {
    "calling": "   s-  s s- --  ",
  },
  Calling {
    "calling": "s - -- s s- --  ",
  },
  Calling {
    "calling": "  s--- s s- --  ",
  },
  Calling {
    "calling": "-  -  -    s--  ",
  },
  Calling {
    "calling": "- s- s-    s--  ",
  },
  Calling {
    "calling": "- ------   s--  ",
  },
  Calling {
    "calling": "-  s- s-   s--  ",
  },
  Calling {
    "calling": "s -- s- -  s--  ",
  },
  Calling {
    "calling": "s  s- ---  s--  ",
  },
  Calling {
    "calling": "  - -----  s--  ",
  },
  Calling {
    "calling": "s s------  s--  ",
  },
  Calling {
    "calling": "s  -  - s  s--  ",
  },
  Calling {
    "calling": "  -  s- s  s--  ",
  },
  Calling {
    "calling": "s s- s- s  s--  ",
  },
  Calling {
    "calling": "-  s-  - s s--  ",
  },
  Calling {
    "calling": "- s--- - s s--  ",
  },
  Calling {
    "calling": "s  s-  --- s--  ",
  },
  Calling {
    "calling": "  - -- --- s--  ",
  },
  Calling {
    "calling": "s s--- --- s--  ",
  },
  Calling {
    "calling": "s -- -  s- s--  ",
  },
  Calling {
    "calling": "  -- s- - ----  ",
  },
  Calling {
    "calling": "   s- --- ----  ",
  },
  Calling {
    "calling": "s - ----- ----  ",
  },
  Calling {
    "calling": "  s------ ----  ",
  },
  Calling {
    "calling": "   -  - s ----  ",
  },
  Calling {
    "calling": "s -  s- s ----  ",
  },
  Calling {
    "calling": "  s- s- s ----  ",
  },
  Calling {
    "calling": "   s-  -------  ",
  },
  Calling {
    "calling": "s - -- -------  ",
  },
  Calling {
    "calling": "  s--- -------  ",
  },
  Calling {
    "calling": "  -- -  s-----  ",
  },
  Calling {
    "calling": "  ---- -- s---  ",
  },
  Calling {
    "calling": "s - s- -- s---  ",
  },
  Calling {
    "calling": "s -  -  s s---  ",
  },
  Calling {
    "calling": "  s- -  s s---  ",
  },
  Calling {
    "calling": "   -  -   - s-  ",
  },
  Calling {
    "calling": "s -  s-   - s-  ",
  },
  Calling {
    "calling": "  s- s-   - s-  ",
  },
  Calling {
    "calling": "  ------  - s-  ",
  },
  Calling {
    "calling": "s - s---  - s-  ",
  },
  Calling {
    "calling": "   s- s-  - s-  ",
  },
  Calling {
    "calling": "   s-  - s- s-  ",
  },
  Calling {
    "calling": "s - -- - s- s-  ",
  },
  Calling {
    "calling": "  s--- - s- s-  ",
  },
  Calling {
    "calling": "   s-    -  - - ",
  },
  Calling {
    "calling": "s - --   -  - - ",
  },
  Calling {
    "calling": "  s---   -  - - ",
  },
  Calling {
    "calling": "  -- - s -  - - ",
  },
  Calling {
    "calling": "  ----- -  s- - ",
  },
  Calling {
    "calling": "s - s-- -  s- - ",
  },
  Calling {
    "calling": "   s- s -  s- - ",
  },
  Calling {
    "calling": "  -- ----  s- - ",
  },
  Calling {
    "calling": "s -  -- s  s- - ",
  },
  Calling {
    "calling": "  s- -- s  s- - ",
  },
  Calling {
    "calling": "   -  s s  s- - ",
  },
  Calling {
    "calling": "  -- - --- s- - ",
  },
  Calling {
    "calling": "   s-   s- s- - ",
  },
  Calling {
    "calling": "s - --  s- s- - ",
  },
  Calling {
    "calling": "  s---  s- s- - ",
  },
  Calling {
    "calling": "s -  --   - --- ",
  },
  Calling {
    "calling": "  s- --   - --- ",
  },
  Calling {
    "calling": "   -  s   - --- ",
  },
  Calling {
    "calling": "  -- s--  - --- ",
  },
  Calling {
    "calling": "  -- - - s- --- ",
  },
  Calling {
    "calling": "  -- s -- ----- ",
  },
  Calling {
    "calling": "   -    s ----- ",
  },
  Calling {
    "calling": "s -  s  s ----- ",
  },
  Calling {
    "calling": "  s- s  s ----- ",
  },
  Calling {
    "calling": "   s-   ------- ",
  },
  Calling {
    "calling": "s - --  ------- ",
  },
  Calling {
    "calling": "  s---  ------- ",
  },
  Calling {
    "calling": "  -- - s------- ",
  },
  Calling {
    "calling": "  ----- - s---- ",
  },
  Calling {
    "calling": "s - s-- - s---- ",
  },
  Calling {
    "calling": "   s- s - s---- ",
  },
  Calling {
    "calling": "  -- ---- s---- ",
  },
  Calling {
    "calling": "s -  -- s s---- ",
  },
  Calling {
    "calling": "  s- -- s s---- ",
  },
  Calling {
    "calling": "   -  s s s---- ",
  },
  Calling {
    "calling": "   -      - s-- ",
  },
  Calling {
    "calling": "s -  s    - s-- ",
  },
  Calling {
    "calling": "  s- s    - s-- ",
  },
  Calling {
    "calling": "  -- s s  - s-- ",
  },
  Calling {
    "calling": "   s-    s- s-- ",
  },
  Calling {
    "calling": "s - --   s- s-- ",
  },
  Calling {
    "calling": "  s---   s- s-- ",
  },
  Calling {
    "calling": "  -- - s s- s-- ",
  },
  Calling {
    "calling": "  -- - - -  - s ",
  },
  Calling {
    "calling": "  -- s --  s- s ",
  },
  Calling {
    "calling": "   -    s  s- s ",
  },
  Calling {
    "calling": "s -  s  s  s- s ",
  },
  Calling {
    "calling": "  s- s  s  s- s ",
  },
  Calling {
    "calling": "   s-   -- s- s ",
  },
  Calling {
    "calling": "s - --  -- s- s ",
  },
  Calling {
    "calling": "  s---  -- s- s ",
  },
  Calling {
    "calling": "  -- - s-- s- s ",
  },
  Calling {
    "calling": "   s--- -  -   -",
  },
  Calling {
    "calling": "  --- s -  -   -",
  },
  Calling {
    "calling": "s - s s -  -   -",
  },
  Calling {
    "calling": "   s ----  -   -",
  },
  Calling {
    "calling": "   - -- s  -   -",
  },
  Calling {
    "calling": "s -   s s  -   -",
  },
  Calling {
    "calling": "  s-  s s  -   -",
  },
  Calling {
    "calling": "   s - --- -   -",
  },
  Calling {
    "calling": "  ---   s- -   -",
  },
  Calling {
    "calling": "s - s   s- -   -",
  },
  Calling {
    "calling": "   ---  s- -   -",
  },
  Calling {
    "calling": "   - --   -  s -",
  },
  Calling {
    "calling": "s -   s   -  s -",
  },
  Calling {
    "calling": "  s-  s   -  s -",
  },
  Calling {
    "calling": "   s s--  -  s -",
  },
  Calling {
    "calling": "   s - - s-  s -",
  },
  Calling {
    "calling": "   s s -- -- s -",
  },
  Calling {
    "calling": "s -     s -- s -",
  },
  Calling {
    "calling": "  s-    s -- s -",
  },
  Calling {
    "calling": "   - s  s -- s -",
  },
  Calling {
    "calling": "  ---   ---- s -",
  },
  Calling {
    "calling": "s - s   ---- s -",
  },
  Calling {
    "calling": "   ---  ---- s -",
  },
  Calling {
    "calling": "   s - s---- s -",
  },
  Calling {
    "calling": "   s--- - s- s -",
  },
  Calling {
    "calling": "  --- s - s- s -",
  },
  Calling {
    "calling": "s - s s - s- s -",
  },
  Calling {
    "calling": "   s ---- s- s -",
  },
  Calling {
    "calling": "   - -- s s- s -",
  },
  Calling {
    "calling": "s -   s s s- s -",
  },
  Calling {
    "calling": "  s-  s s s- s -",
  },
  Calling {
    "calling": "   s s- -  -   s",
  },
  Calling {
    "calling": "  --- ---  -   s",
  },
  Calling {
    "calling": "s - s ---  -   s",
  },
  Calling {
    "calling": "   ------  -   s",
  },
  Calling {
    "calling": "s -   - s  -   s",
  },
  Calling {
    "calling": "  s-  - s  -   s",
  },
  Calling {
    "calling": "   - s- s  -   s",
  },
  Calling {
    "calling": "  ---  --- -   s",
  },
  Calling {
    "calling": "s - s  --- -   s",
  },
  Calling {
    "calling": "   --- --- -   s",
  },
  Calling {
    "calling": "   s -  s- -   s",
  },
  Calling {
    "calling": "   -  -   -  - s",
  },
  Calling {
    "calling": "s -  s-   -  - s",
  },
  Calling {
    "calling": "  s- s-   -  - s",
  },
  Calling {
    "calling": "  ------  -  - s",
  },
  Calling {
    "calling": "s - s---  -  - s",
  },
  Calling {
    "calling": "   s- s-  -  - s",
  },
  Calling {
    "calling": "   s-  - s-  - s",
  },
  Calling {
    "calling": "s - -- - s-  - s",
  },
  Calling {
    "calling": "  s--- - s-  - s",
  },
  Calling {
    "calling": "-  -  -    - - s",
  },
  Calling {
    "calling": "- s- s-    - - s",
  },
  Calling {
    "calling": "- ------   - - s",
  },
  Calling {
    "calling": "-  s- s-   - - s",
  },
  Calling {
    "calling": "s -- s- -  - - s",
  },
  Calling {
    "calling": "s  s- ---  - - s",
  },
  Calling {
    "calling": "  - -----  - - s",
  },
  Calling {
    "calling": "s s------  - - s",
  },
  Calling {
    "calling": "s  -  - s  - - s",
  },
  Calling {
    "calling": "  -  s- s  - - s",
  },
  Calling {
    "calling": "s s- s- s  - - s",
  },
  Calling {
    "calling": "-  s-  - s - - s",
  },
  Calling {
    "calling": "- s--- - s - - s",
  },
  Calling {
    "calling": "s  s-  --- - - s",
  },
  Calling {
    "calling": "  - -- --- - - s",
  },
  Calling {
    "calling": "s s--- --- - - s",
  },
  Calling {
    "calling": "s -- -  s- - - s",
  },
  Calling {
    "calling": "  ---- -- -- - s",
  },
  Calling {
    "calling": "s - s- -- -- - s",
  },
  Calling {
    "calling": "s -  -  s -- - s",
  },
  Calling {
    "calling": "  s- -  s -- - s",
  },
  Calling {
    "calling": "  -- -  ---- - s",
  },
  Calling {
    "calling": "   s-  s---- - s",
  },
  Calling {
    "calling": "s - -- s---- - s",
  },
  Calling {
    "calling": "  s--- s---- - s",
  },
  Calling {
    "calling": "  -- s- - s- - s",
  },
  Calling {
    "calling": "   s- --- s- - s",
  },
  Calling {
    "calling": "s - ----- s- - s",
  },
  Calling {
    "calling": "  s------ s- - s",
  },
  Calling {
    "calling": "   -  - s s- - s",
  },
  Calling {
    "calling": "s -  s- s s- - s",
  },
  Calling {
    "calling": "  s- s- s s- - s",
  },
  Calling {
    "calling": "s -   -   -  s s",
  },
  Calling {
    "calling": "  s-  -   -  s s",
  },
  Calling {
    "calling": "   - s-   -  s s",
  },
  Calling {
    "calling": "   s----  -  s s",
  },
  Calling {
    "calling": "  --- s-  -  s s",
  },
  Calling {
    "calling": "s - s s-  -  s s",
  },
  Calling {
    "calling": "  ---  - s-  s s",
  },
  Calling {
    "calling": "s - s  - s-  s s",
  },
  Calling {
    "calling": "   --- - s-  s s",
  },
  Calling {
    "calling": "   s-- -- -- s s",
  },
  Calling {
    "calling": "   - -  s -- s s",
  },
  Calling {
    "calling": "   s -  ---- s s",
  },
  Calling {
    "calling": "  ---  s---- s s",
  },
  Calling {
    "calling": "s - s  s---- s s",
  },
  Calling {
    "calling": "   --- s---- s s",
  },
  Calling {
    "calling": "   s s- - s- s s",
  },
  Calling {
    "calling": "  --- --- s- s s",
  },
  Calling {
    "calling": "s - s --- s- s s",
  },
  Calling {
    "calling": "   ------ s- s s",
  },
  Calling {
    "calling": "s -   - s s- s s",
  },
  Calling {
    "calling": "  s-  - s s- s s",
  },
  Calling {
    "calling": "   - s- s s- s s",
  },
  Calling {
    "calling": "s -  --   -  ---",
  },
  Calling {
    "calling": "  s- --   -  ---",
  },
  Calling {
    "calling": "   -  s   -  ---",
  },
  Calling {
    "calling": "  -- s--  -  ---",
  },
  Calling {
    "calling": "  -- - - s-  ---",
  },
  Calling {
    "calling": "  -- s -- -- ---",
  },
  Calling {
    "calling": "   -    s -- ---",
  },
  Calling {
    "calling": "s -  s  s -- ---",
  },
  Calling {
    "calling": "  s- s  s -- ---",
  },
  Calling {
    "calling": "   s-   ---- ---",
  },
  Calling {
    "calling": "s - --  ---- ---",
  },
  Calling {
    "calling": "  s---  ---- ---",
  },
  Calling {
    "calling": "  -- - s---- ---",
  },
  Calling {
    "calling": "  ----- - s- ---",
  },
  Calling {
    "calling": "s - s-- - s- ---",
  },
  Calling {
    "calling": "   s- s - s- ---",
  },
  Calling {
    "calling": "  -- ---- s- ---",
  },
  Calling {
    "calling": "s -  -- s s- ---",
  },
  Calling {
    "calling": "  s- -- s s- ---",
  },
  Calling {
    "calling": "   -  s s s- ---",
  },
  Calling {
    "calling": "  ----- -  -  s-",
  },
  Calling {
    "calling": "s - s-- -  -  s-",
  },
  Calling {
    "calling": "   s- s -  -  s-",
  },
  Calling {
    "calling": "  -- ----  -  s-",
  },
  Calling {
    "calling": "s -  -- s  -  s-",
  },
  Calling {
    "calling": "  s- -- s  -  s-",
  },
  Calling {
    "calling": "   -  s s  -  s-",
  },
  Calling {
    "calling": "  -- - --- -  s-",
  },
  Calling {
    "calling": "   s-   s- -  s-",
  },
  Calling {
    "calling": "s - --  s- -  s-",
  },
  Calling {
    "calling": "  s---  s- -  s-",
  },
]
`;
